cwl/src/main/scala/cwl/CommandOutputExpression.scala

```diff
@@ -1,126 +1,49 @@
-        case CommandOutputBinding(Some(glob), _, None) =>
-          GlobEvaluator.globPaths(glob, parameterContext, ioFunctionSet) match {
-            case Seq(fileName) => WomString(fileName).asRight
-            case array => WomArray(WomArrayType(WomStringType), array.map(WomString.apply)).asRight
-          }
-
-
-        case CommandOutputBinding(glob, loadContents, Some(Expression(expression))) =>
-
-          val paths: Seq[String] = glob.toSeq flatMap { globValue =>
-            GlobEvaluator.globPaths(globValue, parameterContext, ioFunctionSet)
-          }
-
-          val _loadContents: Boolean = loadContents getOrElse false
-
-          val womMaps: Array[Map[String, String]] =
-            paths.toArray map {
-              (path: String) =>
-                // TODO: WOM: basename/dirname/size/checksum/etc.
-                val globPathWithDirectory = GlobFunctions.prefixWithGlobDir(path)
-
-                val contents =
-                  Map("contents" -> load64KiB(globPathWithDirectory, ioFunctionSet)).
-                    filter(_ => _loadContents)
-
-                val location = Map("location" -> path)
-
-                location ++ contents
-            }
-
-          val outputEvalParameterContext: ParameterContext = parameterContext.setSelf(womMaps)
-
-          expression.
-            fold(EvaluateExpression).
-            apply(outputEvalParameterContext).
-            cata(Right(_),Left(_)). // this is because toEither is not a thing in scala 2.11.
-            leftMap(e => NonEmptyList(e.getMessage, e.getStackTrace.map(_.toString).toList))
-      }
-    }
-    //To facilitate ECMAScript evaluation, filenames are stored in a map under the key "location"
-    val womValue =
-      outputBindingEvaluationResult map {
-        case WomArray(_, Seq(WomMap(WomMapType(WomStringType, WomStringType), results))) => results(WomString("location"))
-        case other => other
-      }
-
-    //If the value is a string but the output is expecting a file, we consider that string a POSIX "glob" and apply
-    //it accordingly to retrieve the file list to which it expands.
-    val globbedIfFile:ErrorOr[WomValue] =
-      (womValue, cwlExpressionType) match {
-
-        //In the case of a single file being expected, we must enforce that the glob only represents a single file
-        case (Right(WomString(glob)), WomSingleFileType) =>
-          Await.result(ioFunctionSet.glob(glob), Duration.Inf) match {
-            case head :: Nil => WomString(head).validNel
-            case list => s"expecting a single File glob but instead got ${list.toList.mkString(", ")}".invalidNel
-          }
-
-        case (other, _) => other.toValidated
-      }
-
-    //CWL tells us the type this output is expected to be.  Attempt to coerce the actual output into this type.
-    globbedIfFile.toTry.flatMap(cwlExpressionType.coerceRawValue).toErrorOr
+    CommandOutputBinding.generateOutputWomValue(
```

Comment:
I think we are playing tug of war with this code, as I had moved it out of `CommandOutputBinding` to here.  In general I'd like to understand the motivation behind the call to another function instead using this class.

Reply:
The amount of business logic floating around various supposed model classes was getting lengthy and hard to jump between.

I thought about leaving bits and pieces scattered, but wanted it all collected together in one pit-of-despair. So I separated the models and business logic leaving the case classes as simple as possible with either no `def`s at all, or one-liners.

Still due to my lack of naming skills, I didn't have a good place to put all the globbing/output logic so it got chucked into the companion object of the class that tracks [`CommandOutputBinding`](http://www.commonwl.org/v1.0/CommandLineTool.html#CommandOutputBinding)<sup>1</sup> versus [`CommandOutputParameter`](http://www.commonwl.org/v1.0/CommandLineTool.html#CommandOutputParameter)<sup>2</sup>. I'm open to either or even a new `object CwlOutputFigureOuter` since the business logic methods are now static.

---

<sup>1</sup> The CWL globbing behavior is dense. For those who haven't had the pleasure:

> Find files relative to the output directory, using POSIX glob(3) pathname matching. If an array is provided, find files that match any pattern in the array. If an expression is provided, the expression must return a string or an array of strings, which will then be evaluated as one or more glob patterns. Must only match and return files which actually exist. For each file matched in glob, read up to the first 64 KiB of text from the file and place it in the contents field of the file object for manipulation by outputEval. Evaluate an expression to generate the output value. If glob was specified, the value of self must be an array containing file objects that were matched. If no files were matched, self must be a zero length array; if a single file was matched, the value of self is an array of a single element. Additionally, if loadContents is true, the File objects must include up to the first 64 KiB of file contents in the contents field.

<sup>2</sup> _Then_ for each of that result that may be an int, float, string, file, etc.:

> Only valid when type: File or is an array of items: File. Provides a pattern or expression specifying files or directories that must be included alongside the primary file. All listed secondary files must be present. An implementation may fail workflow execution if an expected secondary file does not exist. If the value is an expression, the value of self in the expression must be the primary input or output File object to which this binding applies. The basename, nameroot and nameext fields must be present in self. For CommandLineTool outputs the path field must also be present. The expression must return a filename string relative to the path to the primary File, a File or Directory object with either path or location and basename fields set, or an array consisting of strings or File or Directory objects. It is legal to reference an unchanged File or Directory object taken from input as a secondaryFile. To work on non-filename-preserving storage systems, portable tool descriptions should avoid constructing new values from location, but should construct relative references using basename or nameroot instead. If a value in secondaryFiles is a string that is not an expression, it specifies that the following pattern should be applied to the path of the primary file to yield a filename relative to the primary File: If string begins with one or more caret ^ characters, for each caret, remove the last file extension from the path (the last period . and all following characters). If there are no file extensions, the path is unchanged. Append the remainder of the string to the end of the file path.